File: src/types/connection.ts
/**
 * SSH接続設定
 *
 * AsyncStorageに永続化される接続情報を表す。
 */
export interface Connection {
  /** UUID v4 */
  id: string;
  /** 表示名 (e.g., "Production Server") */
  name: string;
  /** ホスト名 or IPアドレス */
  host: string;
  /** SSHポート (default: 22) */
  port: number;
  /** SSHユーザー名 */
  username: string;
  /** 認証方式 */
  authMethod: 'password' | 'key';
  /** SSH鍵ID (key認証時) */
  keyId?: string;
  /** 接続タイムアウト秒 (default: 30) */
  timeout: number;
  /** Keepalive間隔秒 (default: 60, 0 = 無効) */
  keepAliveInterval: number;

  // メタ情報
  /** カスタムアイコン名 */
  icon?: string;
  /** カード色 (#RRGGBB) */
  color?: string;
  /** タグ */
  tags?: string[];
  /** 最終接続日時 (Unix timestamp ms) */
  lastConnected?: number;
  /** 作成日時 (Unix timestamp ms) */
  createdAt: number;
  /** 更新日時 (Unix timestamp ms) */
  updatedAt: number;
}

/**
 * 接続の作成に必要な情報
 */
export type ConnectionInput = Omit<Connection, 'id' | 'createdAt' | 'updatedAt'>;

/**
 * 接続のランタイム状態
 *
 * 永続化されない、現在の接続状態を表す。
 */
export interface ConnectionState {
  /** 接続ID */
  connectionId: string;
  /** 接続状態 */
  status: 'disconnected' | 'connecting' | 'connected' | 'error';
  /** エラーメッセージ */
  error?: string;
  /** RTT (ms) */
  latency?: number;
  /** 接続開始日時 (Unix timestamp ms) */
  connectedAt?: number;
}

/**
 * デフォルト接続設定
 */
export const DEFAULT_CONNECTION: Partial<Connection> = {
  port: 22,
  timeout: 30,
  keepAliveInterval: 60,
  authMethod: 'password',
};

/**
 * AsyncStorageの保存キー
 */
export const CONNECTIONS_STORAGE_KEY = 'muxpod-connections';

File: flutter/lib/models/connection.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'auth_method.dart';

part 'connection.freezed.dart';
part 'connection.g.dart';

/// SSH接続設定
@freezed
class Connection with _$Connection {
  const factory Connection({
    required String id,
    required String name,
    required String host,
    @Default(22) int port,
    required String username,
    required AuthMethod authMethod,
    String? keyId,
    @Default(30) int timeout,
    @Default(60) int keepAliveInterval,
    String? icon,
    String? color,
    @Default([]) List<String> tags,
    DateTime? lastConnected,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Connection;

  factory Connection.fromJson(Map<String, dynamic> json) =>
      _$ConnectionFromJson(json);
}

/// 接続設定のバリデーション拡張
extension ConnectionValidation on Connection {
  /// バリデーション実行
  List<String> validate() {
    final errors = <String>[];

    if (name.isEmpty || name.length > 50) {
      errors.add('Name must be 1-50 characters');
    }

    if (host.isEmpty) {
      errors.add('Host is required');
    }

    if (port < 1 || port > 65535) {
      errors.add('Port must be 1-65535');
    }

    if (username.isEmpty || username.length > 32) {
      errors.add('Username must be 1-32 characters');
    }

    if (timeout < 5 || timeout > 120) {
      errors.add('Timeout must be 5-120 seconds');
    }

    if (keepAliveInterval != 0 &&
        (keepAliveInterval < 10 || keepAliveInterval > 300)) {
      errors.add('Keep-alive interval must be 0 (off) or 10-300 seconds');
    }

    return errors;
  }

  /// バリデーションが通るか
  bool get isValid => validate().isEmpty;
}
